% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/twas.R
\name{harmonize_twas}
\alias{harmonize_twas}
\title{Function to perform allele flip QC and harmonization on the weights and GWAS against LD for a region.
FIXME: GWAS loading function from Haochen for both tabix & column-mapping yml application}
\usage{
harmonize_twas(twas_weights_data, ld_meta_file_path, gwas_meta_file)
}
\arguments{
\item{twas_weights_data}{List of list of twas weights output from from generate_twas_db function.}

\item{ld_meta_file_path}{A tab-delimited data frame with colname "#chrom", "start", "end", "path", where "path" column
contains file paths for both LD matrix and bim file and is separated by ",". Bim file input would expect no headers, while the
columns are aligned in the order of "chrom", "variants", "GD", "pos", "A1", "A2", "variance", "allele_freq", "n_nomiss".}

\item{gwas_meta_file}{A file path for a dataframe table with column of "study_id", "chrom" (integer), "file_path",
"column_mapping_file". Each file in "file_path" column is tab-delimited dataframe of GWAS summary statistics with column name
"chrom" (or #chrom" if tabix-indexed), "pos", "A2", "A1".}
}
\value{
A list of list for harmonized weights and dataframe of gwas summary statistics that is add to the original input of
twas_weights_data under each context.
}
\description{
Function Conditions:
- processes data in the format of either the output from load_twas_weights/generate_twas_db or
  refined_twas_weights_data from twas pipeline.
- For the first format, we expect there is only one gene's information, that can be accessed through `region_info_obj`
  and refined_twas_weights_data contains per region multiple gene's refined weights data.
}
\details{
Main Steps:
1. allele QC for TWAS weights against the LD meta
2. allele QC for GWA summary stats against the LD meta
3. adjust susie/mvsusie weights based on the overlap variants
}
