---
title: "cis-QTL analysis and GWAS integration workflow"
author: "Anqi Wang and Gao Wang"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{cis-QTL analysis and GWAS integration workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


Here we demonstrate the use of the `pecotmr` package to perform
a version of colocalization, TWAS and MR between a pair of molecular QTL and GWAS.

## Set up your environment

Load the `pecotmr` package.

```{r load-pkgs}
library(pecotmr)
```

## Load data

```{r}
gwas = readRDS("CLU_gwas.rds")
eqtl = readRDS("pseudo_bulk_CLU.ENSG00000120885.rds")
```

```{r}
head(gwas)
```

```{r}
names(eqtl)
```

```{r}
head(eqtl$X)
```

```{r}
head(eqtl$y_res)
```

```{r}
dim(eqtl$X)
```

## Fine-mapping using SuSiE

In FunGen-xQTL we have a [simple pipeline](https://github.com/StatFunGen/xqtl-pipeline/blob/main/code/fine_mapping/SuSiE/SuSiE.ipynb) for this.

```{r}
library(susieR)
library(tidyverse)
get_cs_index <- function(snps_idx, fitted_data) {
    idx <- tryCatch(
        which(
            pmap(list(a = fitted_data$sets$cs), function(a) snps_idx %in% a) %>% unlist()
        ),
        error = function(e) NA_integer_
    )
    if(length(idx) == 0) return(NA_integer_)
    return(idx)
}
signal_cutoff = 0.1
coverage = 0.7
susie_res <- susie(eqtl$X, eqtl$y_res, 
                 L=20,
                 max_iter=500,
                 estimate_residual_variance=TRUE,
                 estimate_prior_variance=TRUE,
                 refine=TRUE,
                 compute_univariate_zscore=FALSE,
                 coverage=coverage)
```

```{r}
secondary_coverage = 0.2
susie_res$sets_secondary = susie_get_cs(susie_res, eqtl$X, coverage=secondary_coverage)
```

```{r}
compute_maf <- function(geno){
  f <- mean(geno,na.rm = TRUE)/2
  return(min(f, 1-f))
}
variants_index = c(which(susie_res$pip >= signal_cutoff), unlist(susie_res$sets$cs)) %>% unique %>% sort
maf = apply(eqtl$X, 2, compute_maf)[variants_index]
all_variants = gsub("_",":",names(susie_res$pip))
variants = gsub("_",":",names(susie_res$pip)[variants_index])
pip = susie_res$pip[variants_index]
cs_info = map_int(variants_index, ~get_cs_index(.x, susie_res))
cs_index = ifelse(is.na(cs_info), 0, str_replace(names(susie_res$sets$cs)[cs_info], "L", "") %>% as.numeric)
univariate_res = univariate_regression(eqtl$X[, variants_index, drop=F], eqtl$y_res)
# here betahat and sebetahat were computed from standardized X and Y
susie_res$top_loci = cbind(variants, maf, univariate_res$betahat, univariate_res$sebetahat, pip, cs_index)
colnames(susie_res$top_loci) = c("variant_id", "maf", "bhat", "sbhat", "pip", "cs_index_primary")#, "cs_index_secondary")
rownames(susie_res$top_loci) = NULL
```

```{r}
susie_res$top_loci # to test you can set it to 1) 0.1 and 0.95 vs 2) 0.1 and 0.2
# parameter: default_coverage = 0.95
# parameter: secondary_coverage = 0.7
# parameter: signal_cutoff = 0.1
```

```{r}
pnorm(-0.122514739342209/0.0249868671221125)
```

PIP of the two SNPs of interest are:

```{r}
susie_res$pip["chr8:27608798_C_T"]
```

```{r}
susie_res$pip["chr8:27637095_G_A"]
```

## TWAS

Use SuSiE,

```{r}
R = cor(eqtl$X)
```

```{r}
devtools::load_all("~/GIT/software/pecotmr")
```

```{r}
weights = susie_weights(susie_res)
```

```{r}
twas_z(weights, gwas$Z, R=R)
```

```{r}
cor(eqtl$X%*%weights, eqtl$y_res)^2
```

Now select only the top loci,

```{r}
plot(weights)
```

```{r}
weights[!all_variants %in% susie_res$top_loci[,1]] = 0
```

```{r}
plot(weights)
```

```{r}
twas_z(weights, gwas$Z, R=R)
```

Use elastic net

```{r}
weights = glmnet_weights(eqtl$X, eqtl$y_res)
cor(eqtl$X%*%weights, eqtl$y_res)^2
twas_z(weights, gwas$Z, R=R)
```

```{r}
plot(weights)
```

Use LASSO,

```{r}
weights = glmnet_weights(eqtl$X, eqtl$y_res,alpha=1)
cor(eqtl$X%*%weights, eqtl$y_res)^2
twas_z(weights, gwas$Z, R=R)
```

```{r}
plot(weights)
```

Use varbvs,

```{r}
library(varbvs)
twas_varbvs = function(X,y,z,R=NULL) {
    if (is.null(R)) R = cor(X)
    weights = varbvs(X,Z=NULL,y,logodds = seq(-3,-1,0.1),verbose=F)$beta
    print(cor(X%*%weights, y)^2)
    twas_z(weights, z, R)
}

twas_varbvs(eqtl$X,eqtl$y_res, gwas$Z, R=R)
```

Use mr.ash

```{r}
lasso.weights = glmnet_weights(eqtl$X, eqtl$y_res,alpha=1)
weights = mrash_weights(eqtl$X, eqtl$y_res, init_prior_sd=TRUE, beta.init=lasso.weights)
cor(eqtl$X%*%weights, eqtl$y_res)^2
twas_z(weights, gwas$Z, R=R)
```

```{r}
plot(weights)
```

```{r}
weights = mrash_weights(eqtl$X, eqtl$y_res, init_prior_sd=FALSE, beta.init=lasso.weights)
cor(eqtl$X%*%weights, eqtl$y_res)^2
twas_z(weights, gwas$Z, R=R)
```

```{r}
plot(weights)
```

Enrichment analysis 

```{r}
gwas_path <- "gwas_toy.rds"
qtl_path <- "QTL_toy.rds"
xqtl_enrichment_wrapper(gwas_files = gwas_path, xqtl_files = qtl_path,
                        xqtl_finemapping_obj = c("Mic", "susie_result_trimmed"), xqtl_varname_obj = c("Mic", "variant_names"))
```
